Include RestRequestOption interface and OnResult implementation

tiagralee opened the pull request:
- add RestRequestOption in restlib
- add unit tests on rest request result
- update arrai transformation

Closes #105 #159

tiagralee commented:
Failed on compare Sysl and arr.ai transforms output, still need to work on the sysl transformation.

andrewemeryanz commented on codegen/arrai/svc_service.arrai:
```diff
@@ -89,10 +89,11 @@ let sysl = //{./sysl};
             if result.HTTPResponse.StatusCode == http.StatusUnauthorized {
                 ${returnDownstreamError("Unauthorized", "result", "nil")}
             }
+            restlib.OnRestRequestResultHTTPResult(result, err, opts)
```
This needs to go straight after `restlib.DoHTTPRequest` so that it always gets called regardless of whether it was an error or not.

andrewemeryanz commented on codegen/tests/cardspb/cards.pb.go:
```diff
@@ -65,9 +65,7 @@ func init() {
 	proto.RegisterType((*Card)(nil), "cards.v1.Card")
 }
 
-func init() {
```
Remove these protobuf changes please. They're not affected by your change. It looks like `go fmt` isn't formatting your output correctly.

andrewemeryanz commented on codegen/transforms/grpc_interface.sysl:
```diff
@@ -183,7 +183,7 @@ CodeGenTransform:
         ParameterDecl = paramDecl("ctx", "context.Context").ParameterDecl
         ParameterDeclC = ["req"] -> <sequence of ParameterDeclC>(ident:
-          ParameterDecl = paramDecl(ident, "*" + package + "." + methodDef(ep).methodName + "Request").ParameterDecl
+          ParameterDecl = paramDecl(ident, "*" + package + "." + methodDef(ep).methodName + "Request" + ", opts ...restlib.RestRequestOption").ParameterDecl
```
This change only affects REST requests, not gRPC requests.

tiagralee replied:
This is the downstream service for gRPC which is the interface for REST service, as below example.
interface
```
// GetStuff Client
type GetStuffClient struct {
	PostStuff func(ctx context.Context, req *simple.PostStuffRequest, opts ...restlib.RestRequestOption) (*simple.Str, error)
}
```
handler
```
// GetStuff ...
func (s *GrpcServiceHandler) GetStuff(ctx context.Context, req *pb.GetStuffRequest) (*pb.GetStuffResponse, error) {
	if s.serviceInterface.GetStuff == nil {
		return s.unimpl.GetStuff(ctx, req)
	}

	ctx, cancel := s.genCallback.DownstreamTimeoutContext(ctx)
	defer cancel()
	client := GetStuffClient{
                //here, the simpleSimpleService.PostStuff has the opts
		PostStuff: s.simpleSimpleService.PostStuff,
	}

	return s.serviceInterface.GetStuff(ctx, req, client)
}
```

andrewemeryanz commented on restlib/restlib_test.go:
```diff
@@ -283,3 +284,26 @@ func TestSendHTTPResponseContentTypeImage(t *testing.T) {
+		spy.MockCallback(result, err)
+	})
+	opts := []RestRequestOption{opt}
+	OnRestRequestResult(restResult, nil, opts)
```
You can inline a bunch of this:
```go
called := false
OnRestRequestResult(restResult, nil, RestRequestOnResult(func(result *core.RestResult, err error) {
    called = true
}))
require.True(t, called)
```

andrewemeryanz reviewed (changes requested):
Minor changes requested.

Commit:
Include RestRequestOption interface and OnResult implementation

Custom handler implementations in some instances require access to
more information than downstream service calls currently permit. This
can include the status code, headers or body in the instance that a
non-successful response was returned.

RestRequestOption instances can be passed into downstream REST
requests to perform custom actions:

response, err := client.GetRestFooBar(ctx, &request,
    restlib.RestRequestOnResult(func(result *restlib.HTTPResult, err error) {
        // access to raw http result and error
    }))

Closes #105 #159
